src/update.rs

```diff
@@ -0,0 +1,599 @@
+use crate::{
+    keys::{
+        OwnershipKeyPair, OwnershipPublicKey, PublishingKeyPair, PublishingPublicKey,
+        RevocationKeyPair, RevocationPublicKey, RevocationSecretKey,
+    },
+    signature::{verify_encsig, verify_sig},
+    transaction::{CommitTransaction, FundingTransaction, SplitTransaction},
+    ChannelBalance,
+};
+use anyhow::{bail, Context};
+use bitcoin::{Amount, Txid};
+use ecdsa_fun::{adaptor::EncryptedSignature, Signature};
+
+/// First message of the channel update protocol.
+pub struct ChannelUpdateProposal {
+    proposed_balance: Balance,
+    time_lock: u32,
+    R: RevocationPublicKey,
+    Y: PublishingPublicKey,
+}
+
+/// Second message of the channel update protocol.
+pub struct ChannelUpdateAccept {
+    R: RevocationPublicKey,
+    Y: PublishingPublicKey,
+}
+
+/// Third message of the channel update protocol.
+pub struct ShareSplitSignature {
+    sig_TX_s: Signature,
+}
+
+/// Fourth message of the channel update protocol.
+pub struct ShareCommitEncryptedSignature {
+    encsig_TX_c: EncryptedSignature,
+}
+
+/// Fifth and last message of the channel update protocol.
+pub struct RevealRevocationSecretKey {
+    r: RevocationSecretKey,
+}
+
+pub struct Party0 {
+    x_self: OwnershipKeyPair,
+    X_other: OwnershipPublicKey,
+    TX_f: FundingTransaction,
+    current_state: ChannelState,
+    revoked_states: Vec<RevokedState>,
+}
+
+pub struct RevokedState {
+    channel_state: ChannelState,
+    r_other: RevocationSecretKey,
+}
+
+pub struct ChannelState {
+    TX_c: CommitTransaction,
+    /// Encrypted signature sent to the counterparty. If the
+    /// counterparty decrypts it with their own `PublishingSecretKey`
+    /// and uses it to sign and broadcast `TX_c`, we will be able to
+    /// extract their `PublishingSecretKey` by using
+    /// `recover_decryption_key`. If said `TX_c` was already revoked,
+    /// we can use it with the `RevocationSecretKey` to punish them.
+    encsig_TX_c_self: EncryptedSignature,
+    /// Encrypted signature received from the counterparty. It can be
+    /// decrypted using our `PublishingSecretkey` and used to sign
+    /// `TX_c`. Keep in mind, that publishing a revoked `TX_c` will
+    /// allow the counterparty to punish us.
+    encsig_TX_c_other: EncryptedSignature,
+    r_self: RevocationKeyPair,
+    R_other: RevocationPublicKey,
+    y_self: PublishingKeyPair,
+    /// Signed split transaction.
+    TX_s: SplitTransaction,
+}
+
+impl Party0 {
+    pub fn propose_channel_update(
+        self,
+        update: ChannelUpdate,
+        time_lock: u32,
+    ) -> anyhow::Result<(Alice1, ChannelUpdateProposal)> {
+        let current_balance = self.balance()?;
+        let proposed_balance = current_balance.apply(update)?;
```

Comment:
Do we need to verify that the proposed update is possible? I.e. the sum of the new outputs still equals the inputs?